core/src/main/java/com/graphhopper/routing/ch/CHAlgoFactoryDecorator.java

```diff
@@ -130,24 +151,42 @@ public CHAlgoFactoryDecorator addWeighting(String weighting) {
     }
 
     public CHAlgoFactoryDecorator addPreparation(PrepareContractionHierarchies pch) {
-        preparations.add(pch);
-        int lastIndex = preparations.size() - 1;
-        if (lastIndex >= weightings.size())
+        // we want to make sure that edge- and node-based preparations are added in the same order as their corresponding
+        // weightings, but changing the order between edge- and node-based preparations is accepted
```

Comment:
Hmm, I know this API is a bit tricky ;) Not sure if it really becomes much better when we enforce having to add the node based preparations before the edge based ones ? If we did it like this we would get rid of calculating the index above, but would have to store the end of the last index of a node-based preparation instead (and throw an error if a node preparation is added after the first edge preparation). If you prefer this option I'll do it otherwise I would leave as is :)